### PR TITLE
Adding permission updater for other device types.

### PR DESCRIPTION
**Pull request by menakaj**
Permission update listener for :
1. Aarduino
2. RaspberryPI
3. Virtual Fire Alarm

### CODE REVIEW

File: components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            if (userStoreManager != null) {
+                if (!userStoreManager.isExistingRole(ArduinoConstants.ROLE_NAME)) {
+                    userStoreManager.addRole(ArduinoConstants.ROLE_NAME, null, getPermissions());
+                } else {
+                    getAuthorizationManager().authorizeRole(ArduinoConstants.ROLE_NAME,
+                            ArduinoConstants.PERM_ENROLL_ARDUINO, CarbonConstants.UI_PERMISSION_ACTION);
+                    getAuthorizationManager().authorizeRole(ArduinoConstants.ROLE_NAME,
+                            ArduinoConstants.PERM_OWNING_DEVICE_VIEW, CarbonConstants.UI_PERMISSION_ACTION);
+                }
+            } } catch (UserStoreException e) {
```
Comment (madhawap) [inline components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java]: Reformat the code here

File: components/device-types/raspberrypi-plugin/org.wso2.carbon.device.mgt.iot.raspberrypi.api/src/main/java/org/wso2/carbon/device/mgt/iot/raspberrypi/service/impl/listener/RaspberryPIPermissionUpdateListener.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            if (userStoreManager != null) {
+                if (!userStoreManager.isExistingRole(RaspberrypiConstants.ROLE_NAME)) {
+                    userStoreManager.addRole(RaspberrypiConstants.ROLE_NAME, null, getPermissions());
+                } else {
+                    getAuthorizationManager().authorizeRole(RaspberrypiConstants.ROLE_NAME,
+                            RaspberrypiConstants.PERM_ENROLL_RASPBERRYPI, CarbonConstants.UI_PERMISSION_ACTION);
+                    getAuthorizationManager().authorizeRole(RaspberrypiConstants.ROLE_NAME,
+                            RaspberrypiConstants.PERM_OWNING_DEVICE_VIEW, CarbonConstants.UI_PERMISSION_ACTION);
+                }
+            } } catch (UserStoreException e) {
```
Comment (madhawap) [inline components/device-types/raspberrypi-plugin/org.wso2.carbon.device.mgt.iot.raspberrypi.api/src/main/java/org/wso2/carbon/device/mgt/iot/raspberrypi/service/impl/listener/RaspberryPIPermissionUpdateListener.java]: Reformat the code here

File: components/device-types/virtual-fire-alarm-plugin/org.wso2.carbon.device.mgt.iot.virtualfirealarm.api/src/main/java/org/wso2/carbon/device/mgt/iot/virtualfirealarm/service/impl/listener/VirtualFireAlarmPermissionUpdateListener.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            if (userStoreManager != null) {
+                if (!userStoreManager.isExistingRole(VirtualFireAlarmConstants.ROLE_NAME)) {
+                    userStoreManager.addRole(VirtualFireAlarmConstants.ROLE_NAME, null, getPermissions());
+                } else {
+                    getAuthorizationManager().authorizeRole(VirtualFireAlarmConstants.ROLE_NAME,
+                            VirtualFireAlarmConstants.PERM_ENROLL_FIRE_ALARM, CarbonConstants.UI_PERMISSION_ACTION);
+                    getAuthorizationManager().authorizeRole(VirtualFireAlarmConstants.ROLE_NAME,
+                            VirtualFireAlarmConstants.PERM_OWNING_DEVICE_VIEW, CarbonConstants.UI_PERMISSION_ACTION);
+                }
+            } } catch (UserStoreException e) {
```
Comment (madhawap) [inline components/device-types/virtual-fire-alarm-plugin/org.wso2.carbon.device.mgt.iot.virtualfirealarm.api/src/main/java/org/wso2/carbon/device/mgt/iot/virtualfirealarm/service/impl/listener/VirtualFireAlarmPermissionUpdateListener.java]: Reformat the code here

Comment (menakaj) [reply components/device-types/virtual-fire-alarm-plugin/org.wso2.carbon.device.mgt.iot.virtualfirealarm.api/src/main/java/org/wso2/carbon/device/mgt/iot/virtualfirealarm/service/impl/listener/VirtualFireAlarmPermissionUpdateListener.java]: Please refer to the Pull Request, 
https://github.com/wso2/carbon-device-mgt-plugins/pull/838

File: components/device-types/raspberrypi-plugin/org.wso2.carbon.device.mgt.iot.raspberrypi.api/src/main/java/org/wso2/carbon/device/mgt/iot/raspberrypi/service/impl/listener/RaspberryPIPermissionUpdateListener.java (hunk 2 above)
Comment (menakaj) [reply components/device-types/raspberrypi-plugin/org.wso2.carbon.device.mgt.iot.raspberrypi.api/src/main/java/org/wso2/carbon/device/mgt/iot/raspberrypi/service/impl/listener/RaspberryPIPermissionUpdateListener.java]: Please refer to the Pull Request,
https://github.com/wso2/carbon-device-mgt-plugins/pull/838

File: components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java (hunk 1 above)
Comment (menakaj) [reply components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java]: Please refer to the pull request,
https://github.com/wso2/carbon-device-mgt-plugins/pull/838

File: components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    @Override
+    public void contextDestroyed(ServletContextEvent servletContextEvent) {
+
+    }
+
+    public static UserStoreManager getUserStoreManager() {
+        RealmService realmService;
+        UserStoreManager userStoreManager;
+        try {
+            PrivilegedCarbonContext ctx = PrivilegedCarbonContext.getThreadLocalCarbonContext();
```
Comment (sinthuja) [inline components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java]: Don't try to lookup for the osgi service each time the method is called. Look up only once, and reuse it. Please change the same for the Authorization Manager as well.

File: components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        } catch (UserStoreException e) {
+            String msg = "Error occurred while retrieving current user store manager";
+            log.error(msg, e);
+            throw new IllegalStateException(msg);
+        }
+        return authorizationManager;
+    }
+
+    private Permission[] getPermissions() {
+
```
Comment (sinthuja) [inline components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java]: Remove unnecessary blank lines.

File: components/device-types/raspberrypi-plugin/org.wso2.carbon.device.mgt.iot.raspberrypi.api/src/main/java/org/wso2/carbon/device/mgt/iot/raspberrypi/service/impl/listener/RaspberryPIPermissionUpdateListener.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            log.error("Error while creating a role and adding a user for Raspberry PI.", e);
+        }
+    }
+
+    @Override
+    public void contextDestroyed(ServletContextEvent servletContextEvent) {
+
+    }
+
+    public static UserStoreManager getUserStoreManager() {
```
Comment (sinthuja) [inline components/device-types/raspberrypi-plugin/org.wso2.carbon.device.mgt.iot.raspberrypi.api/src/main/java/org/wso2/carbon/device/mgt/iot/raspberrypi/service/impl/listener/RaspberryPIPermissionUpdateListener.java]: Please fix it here as well as mentioned in the above comment. https://github.com/wso2/carbon-device-mgt-plugins/pull/540/files#diff-150a2cbabff9215870977105509d8d5dR64

File: components/device-types/virtual-fire-alarm-plugin/org.wso2.carbon.device.mgt.iot.virtualfirealarm.api/src/main/java/org/wso2/carbon/device/mgt/iot/virtualfirealarm/service/impl/listener/VirtualFireAlarmPermissionUpdateListener.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            log.error("Error while creating a role and adding a user for Raspberry PI.", e);
+        }
+    }
+
+    @Override
+    public void contextDestroyed(ServletContextEvent servletContextEvent) {
+
+    }
+
+    public static UserStoreManager getUserStoreManager() {
```
Comment (sinthuja) [inline components/device-types/virtual-fire-alarm-plugin/org.wso2.carbon.device.mgt.iot.virtualfirealarm.api/src/main/java/org/wso2/carbon/device/mgt/iot/virtualfirealarm/service/impl/listener/VirtualFireAlarmPermissionUpdateListener.java]: Please fix it here as well as mentioned in https://github.com/wso2/carbon-device-mgt-plugins/pull/540/files#diff-150a2cbabff9215870977105509d8d5dR64.

File: components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java (hunk 4 above)
Comment (menakaj) [reply components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java]: This is fixed by the commit https://github.com/wso2/carbon-device-mgt-plugins/pull/838/commits/4bed9cb1ffd28ee03f506bb3846521ae2e14d413

File: components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java (hunk 5 above)
Comment (menakaj) [reply components/device-types/arduino-plugin/org.wso2.carbon.device.mgt.iot.arduino.api/src/main/java/org/wso2/carbon/device/mgt/iot/arduino/service/impl/listener/ArduinoPermissionUpdateListener.java]: This is fixed by the commit https://github.com/wso2/carbon-device-mgt-plugins/pull/838/commits/4bed9cb1ffd28ee03f506bb3846521ae2e14d413

File: components/device-types/virtual-fire-alarm-plugin/org.wso2.carbon.device.mgt.iot.virtualfirealarm.api/src/main/java/org/wso2/carbon/device/mgt/iot/virtualfirealarm/service/impl/listener/VirtualFireAlarmPermissionUpdateListener.java (hunk 7 above)
Comment (menakaj) [reply components/device-types/virtual-fire-alarm-plugin/org.wso2.carbon.device.mgt.iot.virtualfirealarm.api/src/main/java/org/wso2/carbon/device/mgt/iot/virtualfirealarm/service/impl/listener/VirtualFireAlarmPermissionUpdateListener.java]: This is fixed by the commit https://github.com/wso2/carbon-device-mgt-plugins/pull/838/commits/4bed9cb1ffd28ee03f506bb3846521ae2e14d413

File: components/device-types/raspberrypi-plugin/org.wso2.carbon.device.mgt.iot.raspberrypi.api/src/main/java/org/wso2/carbon/device/mgt/iot/raspberrypi/service/impl/listener/RaspberryPIPermissionUpdateListener.java (hunk 6 above)
Comment (menakaj) [reply components/device-types/raspberrypi-plugin/org.wso2.carbon.device.mgt.iot.raspberrypi.api/src/main/java/org/wso2/carbon/device/mgt/iot/raspberrypi/service/impl/listener/RaspberryPIPermissionUpdateListener.java]: This is fixed by the commit https://github.com/wso2/carbon-device-mgt-plugins/pull/838/commits/4bed9cb1ffd28ee03f506bb3846521ae2e14d413

**Pull request merged**